Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/VaccinationRepository.kt

```diff
@@ -1,85 +1,207 @@
-    @AppScope private val scope: CoroutineScope,
+    @AppScope private val appScope: CoroutineScope,
+    dispatcherProvider: DispatcherProvider,
+    private val timeStamper: TimeStamper,
+    private val storage: VaccinationStorage,
+    private val valueSetsRepository: ValueSetsRepository,
+    private val vaccinationProofServer: VaccinationProofServer,
 ) {
 
-    private val vc = VaccinationCertificate(
-        firstName = "Max",
-        lastName = "Mustermann",
-        dateOfBirth = LocalDate.now(),
-        vaccinatedAt = Instant.now(),
-        vaccineName = "Comirnaty (mRNA)",
-        vaccineManufacturer = "BioNTech",
-        chargeId = "CB2342",
-        certificateIssuer = "Landratsamt Potsdam",
-        certificateCountry = Country.DE,
-        certificateId = "05930482748454836478695764787840"
-    )
-
-    private val vc1 = VaccinationCertificate(
-        firstName = "Max",
-        lastName = "Mustermann",
-        dateOfBirth = LocalDate.now(),
-        vaccinatedAt = Instant.now(),
-        vaccineName = "Comirnaty (mRNA)",
-        vaccineManufacturer = "BioNTech",
-        chargeId = "CB2342",
-        certificateIssuer = "Landratsamt Potsdam",
-        certificateCountry = Country.DE,
-        certificateId = "05930482748454836478695764787841"
-    )
-
-    private val pc = ProofCertificate(
-        expiresAt = Instant.now()
-    )
-
-    // TODO read from repos
-    val vaccinationInfos: Flow<Set<VaccinatedPerson>> = flowOf(
-        setOf(
-            VaccinatedPerson(
-                setOf(vc),
-                setOf(),
-                isRefreshing = false,
-                lastUpdatedAt = Instant.now()
-            ),
-            VaccinatedPerson(
-                setOf(vc1),
-                setOf(pc),
-                isRefreshing = false,
-                lastUpdatedAt = Instant.now()
-            )
-        )
-    )
+    private val internalData: HotDataFlow<Set<VaccinatedPerson>> = HotDataFlow(
+        loggingTag = TAG,
+        scope = appScope + dispatcherProvider.IO,
+        sharingBehavior = SharingStarted.Lazily,
+    ) {
+        storage.personContainers
+            .map { personContainer ->
+                VaccinatedPerson(
+                    data = personContainer,
+                    valueSet = null,
+                    isUpdatingData = false,
+                    lastError = null
+                )
+            }
+            .toSet()
+            .also { Timber.tag(TAG).v("Restored vaccination data: %s", it) }
+    }
+
+    init {
+        internalData.data
+            .onStart { Timber.tag(TAG).d("Observing test data.") }
+            .onEach { vaccinatedPersons ->
+                Timber.tag(TAG).v("Vaccination data changed: %s", vaccinatedPersons)
+                storage.personContainers = vaccinatedPersons.map { it.data }.toSet()
+            }
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot vaccination data to storage.")
+                throw it
```

Comment:
so this will crash the App then , I would like to change this approach here since you already saw  from CoronaTestRepo we had to do many changes after the feature was already shipped

Reply:
If there is an exception when we serialize the data, it should never have been created.
This is a case of "should never happen", and I don't see a way to recover from this.
If we save it, we corrupt the previously successfully stored data, and "all" is lost.
If we crash the can restart again with a valid data set.